### PR TITLE
feat(Core/Spells): implemented two new custom attributes to handle aura saving rule

### PR DESCRIPTION
**Pull request by sanctum32**
In this PR i implemented two new custom spell attributes and fixed implicit conversation in generic SpellInfo attributes checks

## Changes Proposed:
-  Implemented initial attributes to handle aura saving (SPELL_ATTR0_CU_REJECT_AURA_SAVING, SPELL_ATTR0_CU_FORCE_AURA_SAVING)
-  Applied some code style to old code
-  Added mask stacking rules for SPELL_ATTR0_CU_NEGATIVE_EFF and SPELL_ATTR0_CU_POSITIVE_EFF
-  If Error happens on certain attributes, only attribute which errored will get skipped. This allows other flags to continue loading
-  Raise Ally control aura saving handling been moved to DB and applied SPELL_ATTR0_CU_REJECT_AURA_SAVING to this spell
## Issues Addressed:

- Closes 

## Tests Performed:

- Tested both attributes with "Retribution Aura" (rank 1 but rest of aura ranks should work same too)
- Tested both attributes with "772 Rend" (rank 1 but rest of aura ranks should work same too)

## How to Test the Changes:

Add any aura spell id with one of these attributes and check if aura gets saved or non saved (noting, SPELL_ATTR0_CU_FORCE_AURA_SAVING and SPELL_ATTR0_CU_REJECT_AURA_SAVING doesn't stack, you can pick only one)

## Known Issues and TODO List:

- [ ]
- [ ]

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by Kitzunu** [comment]
What is aura saving?

**Comment by sanctum32** [comment]
@Kitzunu there can be situations where certain aura maybe should be saved but generic check indicates that it can't be saved to characters db. Same happens with spells which maybe shouldn;t saved but same system tells that it can be saved. These attributes can make exceptional forced rules for such behaviors without marking certain spells as negative and similar

**Comment by sanctum32** [comment]
Yet if new collumn is not necessary, i can remake this

**Comment by FrancescoBorzi** [comment]
> I want to preserve possible ports from trinitycore if something may will get ported.

sounds good to me, @azerothcore/testers can anyone double-check this please?

**Comment by Nyeriah** [comment]
The code is good, but I agree with @Kitzunu. While the new column may make it easier for future ports, it breaks normalization as the rows we got there will be filled of lots of 0s and it seems that will be the case most of the time. It's something we tried to avoid countless times @ TC (take for example the several tables that were normalized, creature/creature_addon, quest_template/addon, etc.). 

This is pure tech speech tho, _**in practice**_, the new column would be better as that allows developers to create custom flags as they wish without ever worring about override TC flags.

A solution that doesn't break normalization? We could use an entirely new table `spell_custom_attr_ac`, that way we part away from TrinityCore's implementation while also keeping useless data off the DB.

I found this while looking around: https://github.com/TrinityCore/TrinityCore/commit/b4aa698acb88e1a74bc22d97279148a567b88cad

**Comment by sanctum32** [comment]
@Nyeriah whole spell_custom_attr is more as data holder for optional data, we can always contain null entries there to avoid 0s... either tho, if these 0s are there and they aren't loaded to relative data, it doesn't consumes too much resources (except size on storage). Also we always could use nulls instead. Either tho, duplicating tables would mean to make more trash in database structure (ac structure already is trashed by useless *dbc data)... but yet i would rather add ac prefix into custom attribute collumn

**Comment by sanctum32** [comment]
@Nyeriah i totally wasn't aware that this been added in trinitycore master branch... alright i will update this PR then

File: src/server/game/Spells/SpellInfo.h [hunk 1]
```diff
@@ -323,6 +324,7 @@ class SpellInfo
     uint32 AttributesEx6;
     uint32 AttributesEx7;
     uint32 AttributesCu;
+    uint32 AttributesCu1;
```
Comment (Kitzunu) [inline src/server/game/Spells/SpellInfo.h:327]: not needed

File: src/server/game/Spells/SpellInfo.h [hunk 2]
```diff
@@ -201,6 +201,7 @@ enum SpellCustomAttributes
     SPELL_ATTR0_CU_POSITIVE_EFF0                 = 0x02000000,
     SPELL_ATTR0_CU_POSITIVE_EFF1                 = 0x04000000,
     SPELL_ATTR0_CU_POSITIVE_EFF2                 = 0x08000000,
+    SPELL_ATTR0_CU_REJECT_AURA_SAVING            = 0x10000000,
```
Comment (Kitzunu) [inline src/server/game/Spells/SpellInfo.h:204]: I would prefer to have them together. So pls move SPELL_ATTR0_CU_FORCE_AURA_SAVING to be at 0x20000000

**Review by Kitzunu — CHANGES_REQUESTED** [review]

File: src/server/game/Spells/SpellInfo.h (hunk 1 above)
Comment (UltraNix) [reply src/server/game/Spells/SpellInfo.h:327]: Not needed - not used in core.

File: src/server/game/Spells/SpellInfo.cpp [hunk 3]
```diff
@@ -783,6 +783,7 @@ SpellInfo::SpellInfo(SpellEntry const* spellEntry)
     AttributesEx6 = spellEntry->AttributesEx6;
     AttributesEx7 = spellEntry->AttributesEx7;
     AttributesCu = 0;
+    AttributesCu1 = 0;
```
Comment (UltraNix) [inline src/server/game/Spells/SpellInfo.cpp:786]: Not need  - not used by core.

Comment (sanctum32) [reply src/server/game/Spells/SpellInfo.cpp]: removed

File: src/server/game/Spells/SpellInfo.h (hunk 1 above)
Comment (sanctum32) [reply src/server/game/Spells/SpellInfo.h]: Removed

File: src/server/game/Spells/SpellInfo.h (hunk 2 above)
Comment (sanctum32) [reply src/server/game/Spells/SpellInfo.h:204]: Applied

**Pull request merged**